### PR TITLE
Better validate partial measurement output

### PR DESCRIPTION
**Pull request by efernandez**
* Show source if validation fails
  * Also changes from throwing/crashing to `ROS_ERROR`.

* Add precision arg for covariance validation. This allows to:
    * Relax the default precision when validating the covariance matrix is
      symmetric.
    * Print the covariance matrix with `Eigen::FullPrecision` when the
      symmetry test fails with `isApprox`, so we can see the magnitude of
      the error.

### CODE REVIEW

File: fuse_models/include/fuse_models/common/sensor_proc.h
Comment (svwilliams) [inline fuse_models/include/fuse_models/common/sensor_proc.h:496]: I feel like if the validation fails, we should not be creating a constraint from it. In the same way that tf errors prevent the creation of a constraint. Thoughts?

```suggestion
    return false;
  }
```

Comment (efernandez) [reply fuse_models/include/fuse_models/common/sensor_proc.h:496]: The reason I still create constraints it's because I'm playing with https://github.com/clearpathrobotics/fuse/pull/7 and I still get cases where the covariance is slightly non-symmetric or non-PSD, but the constraint with those slightly bad covariances don't make the optimizer explode.

Maybe we could wait til EOW to see if I get things to actually pass this checks. I'm also planning to create an upstream PR for https://github.com/clearpathrobotics/fuse/pull/7 once I get things to work better.

Comment (ayrton04) [reply fuse_models/include/fuse_models/common/sensor_proc.h:496]: Would an extra parameter work? Something like `permit_invalid`?

**Review by ayrton04 — APPROVED** [review]

File: fuse_models/include/fuse_models/common/sensor_proc.h
Comment (efernandez) [reply fuse_models/include/fuse_models/common/sensor_proc.h:496]: Actually, we could have a parameter to disable the checks. Something like `disable_checks` in Ceres solver.

The checks could be enabled by default, so the user can easily detect issues with the inputs and understand what's wrong with them. This would incur a small overhead.

Once things are known to work and be correct consistently, then `disable_checks: true` could be set, so we don't incur any overhead. If you're ok with something like this, I can add it.

Comment (efernandez) [reply fuse_models/include/fuse_models/common/sensor_proc.h:496]: @svwilliams What's your opinion on this?

File: fuse_models/include/fuse_models/common/sensor_proc.h
Comment (efernandez) [inline fuse_models/include/fuse_models/common/sensor_proc.h]: I'd like to change this to `fuse_core::to_string(covariance_partial, Eigen::FullPrecision)` as well.

And also print the `solver.eigenvalues()`.

Sounds good?

Mostly a heads up, so you wait for those changes before you merge this PR. :smiley:

File: fuse_models/include/fuse_models/common/sensor_proc.h
Comment (svwilliams) [reply fuse_models/include/fuse_models/common/sensor_proc.h:496]: I like the "disable_checks" theory. With checks enabled, failed tests prevent constraints from being generated. And with checks disabled, the constraints are obviously generated all the time.

File: fuse_models/include/fuse_models/common/sensor_proc.h
Comment (svwilliams) [reply fuse_models/include/fuse_models/common/sensor_proc.h]: More information on failure is better than less. Go for it.

Comment (efernandez) [reply fuse_models/include/fuse_models/common/sensor_proc.h]: Added the additional information.

File: fuse_models/include/fuse_models/common/sensor_proc.h
Comment (efernandez) [reply fuse_models/include/fuse_models/common/sensor_proc.h:496]: Add `disable_checks` param (defaults to `false`) to all sensor models and `validate` argument to all `sensor_proc.h` processing functions.

If `disable_checks == false`, then `validate == true` and all validation checks are evaluated. When a validation check fails, no constraint is created for the partial measurements involved.

If `disable_checks == true`, then `validate == false` and no validation check is evaluated. Therefore, constraints are created for all partial measurements.

**Review by svwilliams — APPROVED** [review]

**Pull request merged**
